{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "76729d02-d73a-4eb6-be0f-82b4919ae546",
   "metadata": {},
   "outputs": [],
   "source": [
    "verbose = True\n",
    "import openai\n",
    "openai.api_key=\"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\"\n",
    "\n",
    "import win32clipboard\n",
    "from bs4 import BeautifulSoup\n",
    "from tkinter import Tk \n",
    "import re\n",
    "import requests\n",
    "import io\n",
    "from PyPDF4 import PdfFileReader\n",
    "import numpy as np\n",
    "from tiktoken import encoding_for_model\n",
    "from sentence_transformers import SentenceTransformer\n",
    "import torch.nn.functional as F\n",
    "\n",
    "### UTILITIES\n",
    "def get_clipboard_html():\n",
    "    win32clipboard.OpenClipboard()\n",
    "    try:\n",
    "        html = win32clipboard.GetClipboardData(win32clipboard.RegisterClipboardFormat(\"HTML Format\"))\n",
    "    except TypeError:\n",
    "        html = None\n",
    "    win32clipboard.CloseClipboard()\n",
    "    return html\n",
    "    \n",
    "def get_text_and_source_url_from_clipboard():\n",
    "    text = Tk().clipboard_get()\n",
    "    pattern = re.compile(r'\\[\\d+\\]')\n",
    "    output_string = pattern.sub('', text)\n",
    "\n",
    "    html_content = get_clipboard_html()\n",
    "    if html_content:\n",
    "        soup = BeautifulSoup(html_content, 'html.parser')\n",
    "        a_tags = soup.find_all('a')\n",
    "        # Get the last <a> tag\n",
    "        last_a_tag = a_tags[-1] if a_tags else None\n",
    "        # Extract href attribute\n",
    "        href = last_a_tag.get('href') if last_a_tag else None\n",
    "    else:\n",
    "        return output_string, None\n",
    "    return output_string, href\n",
    "\n",
    "def split_url(url):\n",
    "    # Split the URL at the '#' character\n",
    "    parts = url.split(\"#\")\n",
    "    # Check if URL contains '#'\n",
    "    if len(parts) > 1:\n",
    "        base_url = parts[0]\n",
    "        fragment = parts[1]\n",
    "        return base_url, fragment\n",
    "    else:\n",
    "        print(\"The URL does not contain a # fragment.\")\n",
    "        return url, \"\"\n",
    "\n",
    "def get_first_https_link(full_url):\n",
    "    if full_url is None: return None\n",
    "    url, element_id = split_url(full_url)\n",
    "    try:\n",
    "        # Send a request to the URL\n",
    "        response = requests.get(url)\n",
    "        response.raise_for_status()  # Raise an exception for HTTP errors\n",
    "\n",
    "        # Initialize the BeautifulSoup object with the response content\n",
    "        soup = BeautifulSoup(response.content, 'html.parser')\n",
    "\n",
    "        # Find the element by its ID\n",
    "        li_element = soup.find(id=element_id)\n",
    "\n",
    "        if li_element is None:\n",
    "            print(f\"Element with id '{element_id}' not found on the page.\")\n",
    "            return None\n",
    "\n",
    "        # Find all the 'a' tags in the 'li' element\n",
    "        a_tags = li_element.find_all('a', href=True)\n",
    "\n",
    "        # Look for the first 'a' tag where the href attribute starts with \"https://\"\n",
    "        for a_tag in a_tags:\n",
    "            href = a_tag.get('href')\n",
    "            if href.startswith('http'):\n",
    "                return href\n",
    "\n",
    "        print(f\"No https link found in element with id '{element_id}'.\")\n",
    "    \n",
    "    except requests.RequestException as e:\n",
    "        print(f\"Request failed: {e}\")\n",
    "\n",
    "def scrape_visible_text(url):\n",
    "    if url is None: return None\n",
    "    if url.endswith(\".pdf\"):\n",
    "        print(\"The source is a PDF document.\")\n",
    "        return scrape_pdf_text(url)\n",
    "    else:\n",
    "        try:\n",
    "            # Send a GET request\n",
    "            response = requests.get(url)\n",
    "            response.raise_for_status()  # Raise an exception for HTTP errors\n",
    "            #print(response)\n",
    "            # Initialize BeautifulSoup\n",
    "            soup = BeautifulSoup(response.text, 'html.parser')\n",
    "    \n",
    "            # Remove script and style elements\n",
    "            for script_or_style in soup([\"script\", \"style\"]):\n",
    "                script_or_style.decompose()\n",
    "    \n",
    "            # Get the page text\n",
    "            text = soup.get_text()\n",
    "            \n",
    "            # Break into lines and remove leading/trailing whitespace on each\n",
    "            lines = (line.strip() for line in text.splitlines())\n",
    "            # Break multi-headlines into separate lines\n",
    "            chunks = (phrase.strip() for line in lines for phrase in line.split(\"  \"))\n",
    "            # Drop blank lines\n",
    "            text = '\\n'.join(chunk for chunk in chunks if chunk)\n",
    "    \n",
    "            return text\n",
    "    \n",
    "        except requests.RequestException as e:\n",
    "            print(f\"Request failed: {e}\")\n",
    "            return None\n",
    "            \n",
    "# Function to scrape text\n",
    "def scrape_pdf_text(url):\n",
    "    # Send an HTTP request to the URL of the PDF\n",
    "    response = requests.get(url)\n",
    "\n",
    "    # Check if the request was successful\n",
    "    if response.status_code == 200:\n",
    "        # Get the content of the response in binary form\n",
    "        pdf_content = io.BytesIO(response.content)\n",
    "\n",
    "        # Create a PDF reader object\n",
    "        pdf_reader = PdfFileReader(pdf_content)\n",
    "\n",
    "        # Number of pages in PDF\n",
    "        num_pages = pdf_reader.numPages\n",
    "\n",
    "        # A list to store text from each page of the PDF\n",
    "        text_content = []\n",
    "\n",
    "        # Loop through all pages\n",
    "        for i in range(num_pages):\n",
    "            # Extract text from each page\n",
    "            text = pdf_reader.getPage(i).extractText()\n",
    "            text_content.append(text)\n",
    "\n",
    "        return \" \".join(text_content)\n",
    "    else:\n",
    "        print(\"Failed to retrieve the PDF at:\", url)\n",
    "        return None\n",
    "        \n",
    "def get_claim_and_source_from_clipboard():\n",
    "    text, citation_ref = get_text_and_source_url_from_clipboard()\n",
    "    external_link = get_first_https_link(citation_ref)\n",
    "    print(\"copied text: \" + text)\n",
    "    print(\"source link: \" +external_link)\n",
    "    \n",
    "    source_text = scrape_visible_text(external_link)\n",
    "    print(\"source text found: \\n\"+source_text)\n",
    "    if source_text is None:\n",
    "        source_text=input()\n",
    "    return text, source_text"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "31dd40ba-9229-49d7-8f38-46eff454a8cf",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "copied text: He currently holds the title of Associate Adjunct Professor in Neurosciences at the University of California San Diego.\n",
      "source link: https://profiles.ucsd.edu/nathan.shaner\n",
      "source text found: \n",
      "Nathan Shaner | UCSD Profiles\n",
      "Powered by Altman Clinical and Translational Research Institute\n",
      "UCSD Profiles\n",
      "search, discover, network\n",
      "Home\n",
      "About\n",
      "Help/FAQs\n",
      "History (1)Shaner, NathanSee All Pages\n",
      "Search Options\n",
      "UC Davis\n",
      "UCI\n",
      "UCLA\n",
      "UCSD\n",
      "UCSF\n",
      "All UC Health\n",
      "USC\n",
      "All\n",
      "Everything\n",
      "Research\n",
      "Concepts\n",
      "Awards\n",
      "People\n",
      "UC Health People\n",
      "UC Davis People\n",
      "UCI People\n",
      "UCLA People\n",
      "UCSD People\n",
      "UCSF People\n",
      "USC People\n",
      "Find PeopleFind Everything\n",
      "Login to edit your profile (add a photo, awards, links to other websites, etc.)\n",
      "Edit My Profile\n",
      "My Person List (0)\n",
      "Return to Top\n",
      "Nathan Shaner\n",
      "Title(s)Associate Adjunct Professor, NeurosciencesSchoolVc-health Sciences-schoolsAddress9500 Gilman Drive #La Jolla CA 92093vCardDownload vCard\n",
      "Research  Research Activities and FundingHighly specific control of neurons with photoswitchable bioluminescent optogenetics.NIH R21EY030716Aug 1, 2019 - Jul 31, 2021Role: Principal InvestigatorProbes for Luminescence-based Superresolution MicroscopyNIH R01GM121944Feb 16, 2017 - Jan 31, 2021Role: Principal InvestigatorBioLuminescent OptoGenetics (BL-OG): A Novel and Versatile Strategy for NeuromodulationNIH U01NS099709Sep 30, 2016 - Jun 30, 2020Role: Co-Principal InvestigatorFluorescent proteins for superresolution imagingNIH R01GM109984Sep 30, 2014 - Aug 31, 2020Role: Principal Investigator\n",
      "Bibliographic  Publications\n",
      "Publications listed below are automatically derived from MEDLINE/PubMed and other sources, which might result in incorrect or missing publications.\n",
      "Researchers can login to make corrections and additions, or contact us for help.\n",
      "to make corrections and additions.\n",
      "Newest\n",
      "|\n",
      "Oldest\n",
      "|\n",
      "Most Cited\n",
      "|\n",
      "Most Discussed\n",
      "|\n",
      "Timeline\n",
      "|\n",
      "Field Summary\n",
      "|\n",
      "Plain Text\n",
      "Altmetrics Details\n",
      "PMC Citations indicate the number of times the publication was cited by articles in PubMed Central, and the Altmetric score represents citations in news articles and social media.\n",
      "(Note that publications are often cited in additional ways that are not shown here.)\n",
      "Fields are based on how the National Library of Medicine (NLM) classifies the publication's journal and might not represent the specific topic of the publication.\n",
      "Translation tags are based on the publication type and the MeSH terms NLM assigns to the publication.\n",
      "Some publications (especially newer ones and publications not in PubMed) might not yet be assigned Field or Translation tags.)\n",
      "Click a Field or Translation tag to filter the publications.\n",
      "Bioluminescent Genetically Encoded Glutamate Indicators for Molecular Imaging of Neuronal Activity. ACS Synth Biol. 2023 08 18; 12(8):2301-2309.\n",
      "Petersen ED, Lapan AP, Castellanos Franco EA, Fillion AJ, Crespo EL, Lambert GG, Grady CJ, Zanca AT, Orcutt R, Hochgeschwender U, Shaner NC, Gilad AA.\n",
      "PMID: 37450884; PMCID: PMC10443529.\n",
      "View in: PubMed   Mentions: 1     Fields: Bio BiotechnologyMol Molecular Biology   Translation:Humans\n",
      "A Bioluminescent Activity Dependent (BLADe) Platform for Converting Neuronal Activity to Photoreceptor Activation. bioRxiv. 2023 Jun 27.\n",
      "Crespo EL, Pal A, Prakash M, Silvagnoli AD, Zaidi Z, Gomez-Ramirez M, Tree MO, Shaner NC, Lipscombe D, Moore C, Hochgeschwender U.\n",
      "PMID: 37425742; PMCID: PMC10327117.\n",
      "View in: PubMed   Mentions:\n",
      "CaBLAM! A high-contrast bioluminescent Ca 2+ indicator derived from an engineered Oplophorus gracilirostris luciferase. bioRxiv. 2023 Jun 26.\n",
      "Lambert GG, Crespo EL, Murphy J, Boassa D, Luong S, Hu J, Sprecher B, Tree MO, Orcutt R, Heydari D, Bell AB, Torreblanca Zanca A, Hakimi A, Lipscombe D, Moore CI, Hochgeschwender U, Shaner NC.\n",
      "PMID: 37425712; PMCID: PMC10327125.\n",
      "View in: PubMed   Mentions:\n",
      "A FRET-based excitatory luminopsin with improved efficacy of bioluminescence-driven channelrhodopsin activation. bioRxiv. 2023 Jun 26.\n",
      "Björefeldt A, Murphy J, Crespo EL, Prakash M, Friedman N, Brown TM, Lipscombe D, Moore CI, Hochgeschwender U, Shaner NC.\n",
      "PMID: 37425735; PMCID: PMC10327108.\n",
      "View in: PubMed   Mentions:\n",
      "Selective control of synaptically-connected circuit elements by all-optical synapses. Commun Biol. 2022 01 11; 5(1):33.\n",
      "Prakash M, Murphy J, St Laurent R, Friedman N, Crespo EL, Bjorefeldt A, Pal A, Bhagat Y, Kauer JA, Shaner NC, Lipscombe D, Moore CI, Hochgeschwender U.\n",
      "PMID: 35017641; PMCID: PMC8752598.\n",
      "View in: PubMed   Mentions: 7  Translation:AnimalsCells\n",
      "Endoplasmic reticulum tubules limit the size of misfolded protein condensates. Elife. 2021 09 01; 10.\n",
      "Parashar S, Chidambaram R, Chen S, Liem CR, Griffis E, Lambert GG, Shaner NC, Wortham M, Hay JC, Ferro-Novick S.\n",
      "PMID: 34467852; PMCID: PMC8486381.\n",
      "View in: PubMed   Mentions: 10     Fields: Bio Biology   Translation:HumansAnimalsCells\n",
      "Bioluminescent optogenetic (BL-OG) activation of neurons during mouse postnatal brain development. STAR Protoc. 2021 09 17; 2(3):100667.\n",
      "Crespo EL, Prakash M, Bjorefeldt A, Medendorp WE, Shaner NC, Lipscombe D, Moore CI, Hochgeschwender U.\n",
      "PMID: 34286295; PMCID: PMC8273415.\n",
      "View in: PubMed   Mentions: 2     Fields: Bio BiologySci Science   Translation:AnimalsCells\n",
      "Local Electric Field Controls Fluorescence Quantum Yield of Red and Far-Red Fluorescent Proteins. Front Mol Biosci. 2021; 8:633217.\n",
      "Drobizhev M, Molina RS, Callis PR, Scott JN, Lambert GG, Salih A, Shaner NC, Hughes TE.\n",
      "PMID: 33763453; PMCID: PMC7983054.\n",
      "View in: PubMed   Mentions: 9\n",
      "Aequorea's secrets revealed: New fluorescent proteins with unique properties for bioimaging and biosensing. PLoS Biol. 2020 11; 18(11):e3000936.\n",
      "Lambert GG, Depernet H, Gotthard G, Schultz DT, Navizet I, Lambert T, Adams SR, Torreblanca-Zanca A, Chu M, Bindels DS, Levesque V, Nero Moffatt J, Salih A, Royant A, Shaner NC.\n",
      "PMID: 33137097; PMCID: PMC7660908.\n",
      "View in: PubMed   Mentions: 15     Fields: Bio Biology   Translation:Animals\n",
      "Miniaturized Devices for Bioluminescence Imaging in Freely Behaving Animals. Annu Int Conf IEEE Eng Med Biol Soc. 2020 07; 2020:4385-4389.\n",
      "Celinskis D, Friedman N, Koksharov M, Murphy J, Gomez-Ramirez M, Borton D, Shaner N, Hochgeschwender U, Lipscombe D, Moore C.\n",
      "PMID: 33018967.\n",
      "View in: PubMed   Mentions: 2     Fields: Bio Biomedical Engineering   Translation:AnimalsCells\n",
      "Alkynamide phthalazinones as a new class of TbrPDEB1 inhibitors (Part 2). Bioorg Med Chem. 2019 09 15; 27(18):4013-4029.\n",
      "de Heuvel E, Singh AK, Boronat P, Kooistra AJ, van der Meer T, Sadek P, Blaazer AR, Shaner NC, Bindels DS, Caljon G, Maes L, Sterk GJ, Siderius M, Oberholzer M, de Esch IJP, Brown DG, Leurs R.\n",
      "PMID: 31378593.\n",
      "View in: PubMed   Mentions: 6     Fields: Bio BiochemistryChe Chemistry   Translation:Humans\n",
      "Assay for Detecting Gαi-Mediated Decreases in cAMP in Living Cells. SLAS Discov. 2018 10; 23(9):898-906.\n",
      "Tewson P, Martinka S, Shaner N, Berlot C, Quinn AM, Hughes T.\n",
      "PMID: 29991302.\n",
      "View in: PubMed   Mentions: 6     Fields: Mol Molecular Biology   Translation:HumansAnimalsCells\n",
      "Spatiotemporal manipulation of ciliary glutamylation reveals its roles in intraciliary trafficking and Hedgehog signaling. Nat Commun. 2018 04 30; 9(1):1732.\n",
      "Hong SR, Wang CL, Huang YS, Chang YC, Chang YC, Pusapati GV, Lin CY, Hsu N, Cheng HC, Chiang YC, Huang WE, Shaner NC, Rohatgi R, Inoue T, Lin YC.\n",
      "PMID: 29712905; PMCID: PMC5928066.\n",
      "View in: PubMed   Mentions: 33     Fields: Bio BiologySci Science   Translation:AnimalsCells\n",
      "Characterization of a spectrally diverse set of fluorescent proteins as FRET acceptors for mTurquoise2. Sci Rep. 2017 09 20; 7(1):11999.\n",
      "Mastop M, Bindels DS, Shaner NC, Postma M, Gadella TWJ, Goedhart J.\n",
      "PMID: 28931898; PMCID: PMC5607329.\n",
      "View in: PubMed   Mentions: 35     Fields: Sci Science   Translation:HumansAnimalsCells\n",
      "Blue-Shifted Green Fluorescent Protein Homologues Are Brighter than Enhanced Green Fluorescent Protein under Two-Photon Excitation. J Phys Chem Lett. 2017 Jun 15; 8(12):2548-2554.\n",
      "Molina RS, Tran TM, Campbell RE, Lambert GG, Salih A, Shaner NC, Hughes TE, Drobizhev M.\n",
      "PMID: 28530831; PMCID: PMC5474692.\n",
      "View in: PubMed   Mentions: 16     Fields: Che Chemistry\n",
      "Engineering of mCherry variants with long Stokes shift, red-shifted fluorescence, and low cytotoxicity. PLoS One. 2017; 12(2):e0171257.\n",
      "Shen Y, Chen Y, Wu J, Shaner NC, Campbell RE.\n",
      "PMID: 28241009; PMCID: PMC5328254.\n",
      "View in: PubMed   Mentions: 28     Fields: Med Medicine (General)Sci Science   Translation:AnimalsCells\n",
      "Structural analysis of the bright monomeric yellow-green fluorescent protein mNeonGreen obtained by directed evolution. Acta Crystallogr D Struct Biol. 2016 12 01; 72(Pt 12):1298-1307.\n",
      "Clavel D, Gotthard G, von Stetten D, De Sanctis D, Pasquier H, Lambert GG, Shaner NC, Royant A.\n",
      "PMID: 27917830; PMCID: PMC5137226.\n",
      "View in: PubMed   Mentions: 15  Translation:AnimalsCells\n",
      "New DAG and cAMP Sensors Optimized for Live-Cell Assays in Automated Laboratories. J Biomol Screen. 2016 Mar; 21(3):298-305.\n",
      "Tewson PH, Martinka S, Shaner NC, Hughes TE, Quinn AM.\n",
      "PMID: 26657040; PMCID: PMC4766961.\n",
      "View in: PubMed   Mentions: 53     Fields: Mol Molecular Biology   Translation:HumansCells\n",
      "Special Section Guest Editorial: Protein Photonics for Imaging, Sensing, and Manipulation: Honoring Prof. Osamu Shimomura, a Pioneer of Photonics for Biomedical Research. J Biomed Opt. 2015 Oct; 20(10):101201.\n",
      "Fujita K, Nagai T, Shaner N, Egner A.\n",
      "PMID: 26527157.\n",
      "View in: PubMed   Mentions:    Fields: Bio Biomedical EngineeringOph Ophthalmology   Translation:HumansAnimals\n",
      "Occurrence of Isopenicillin-N-Synthase Homologs in Bioluminescent Ctenophores and Implications for Coelenterazine Biosynthesis. PLoS One. 2015; 10(6):e0128742.\n",
      "Francis WR, Shaner NC, Christianson LM, Powers ML, Haddock SH.\n",
      "PMID: 26125183; PMCID: PMC4488382.\n",
      "View in: PubMed   Mentions: 11     Fields: Med Medicine (General)Sci Science   Translation:AnimalsCells\n",
      "Fluorescent proteins for quantitative microscopy: important properties and practical evaluation. Methods Cell Biol. 2014; 123:95-111.\n",
      "Shaner NC.\n",
      "PMID: 24974024.\n",
      "View in: PubMed   Mentions: 13     Fields: Cel Cell Biology   Translation:HumansAnimals\n",
      "A bright monomeric green fluorescent protein derived from Branchiostoma lanceolatum. Nat Methods. 2013 May; 10(5):407-9.\n",
      "Shaner NC, Lambert GG, Chammas A, Ni Y, Cranfill PJ, Baird MA, Sell BR, Allen JR, Day RN, Israelsson M, Davidson MW, Wang J.\n",
      "PMID: 23524392; PMCID: PMC3811051.\n",
      "View in: PubMed   Mentions: 568     Fields: Cli Clinical Laboratory Techniques   Translation:Animals\n",
      "A comparison across non-model animals suggests an optimal sequencing depth for de novo transcriptome assembly. BMC Genomics. 2013 Mar 12; 14:167.\n",
      "Francis WR, Christianson LM, Kiko R, Powers ML, Shaner NC, Haddock SH.\n",
      "PMID: 23496952; PMCID: PMC3655071.\n",
      "View in: PubMed   Mentions: 48     Fields: Gen Genetics   Translation:AnimalsCells\n",
      "Expression and characterization of the calcium-activated photoprotein from the ctenophore Bathocyroe fosteri: insights into light-sensitive photoproteins. Biochem Biophys Res Commun. 2013 Feb 08; 431(2):360-6.\n",
      "Powers ML, McDermott AG, Shaner NC, Haddock SH.\n",
      "PMID: 23262181; PMCID: PMC3570696.\n",
      "View in: PubMed   Mentions: 3     Fields: Bio BiochemistryBio Biophysics   Translation:AnimalsCells\n",
      "The mFruit collection of monomeric fluorescent proteins. Clin Chem. 2013 Feb; 59(2):440-1.\n",
      "Shaner NC.\n",
      "PMID: 23213078.\n",
      "View in: PubMed   Mentions: 1     Fields: Che Chemistry   Translation:Animals\n",
      "Concurrent imaging of synaptic vesicle recycling and calcium dynamics. Front Mol Neurosci. 2011; 4:34.\n",
      "Li H, Foss SM, Dobryy YL, Park CK, Hires SA, Shaner NC, Tsien RY, Osborne LC, Voglmaier SM.\n",
      "PMID: 22065946; PMCID: PMC3206542.\n",
      "View in: PubMed   Mentions: 38\n",
      "Differential gene expression in the siphonophore Nanomia bijuga (Cnidaria) assessed with multiple next-generation sequencing workflows. PLoS One. 2011; 6(7):e22953.\n",
      "Siebert S, Robinson MD, Tintori SC, Goetz F, Helm RR, Smith SA, Shaner N, Haddock SH, Dunn CW.\n",
      "PMID: 21829563; PMCID: PMC3146525.\n",
      "View in: PubMed   Mentions: 23     Fields: Med Medicine (General)Sci Science   Translation:Animals\n",
      "The maternal and early embryonic transcriptome of the milkweed bug Oncopeltus fasciatus. BMC Genomics. 2011 Jan 25; 12:61.\n",
      "Ewen-Campen B, Shaner N, Panfilio KA, Suzuki Y, Roth S, Extavour CG.\n",
      "PMID: 21266083; PMCID: PMC3040728.\n",
      "View in: PubMed   Mentions: 75     Fields: Gen Genetics   Translation:Animals\n",
      "A monomeric photoconvertible fluorescent protein for imaging of dynamic protein localization. J Mol Biol. 2010 Sep 03; 401(5):776-91.\n",
      "Hoi H, Shaner NC, Davidson MW, Cairo CW, Wang J, Campbell RE.\n",
      "PMID: 20603133.\n",
      "View in: PubMed   Mentions: 29     Fields: Mol Molecular Biology   Translation:Cells\n",
      "Simultaneous visualization of protumorigenic Src and MT1-MMP activities with fluorescence resonance energy transfer. Cancer Res. 2010 Mar 15; 70(6):2204-12.\n",
      "Ouyang M, Huang H, Shaner NC, Remacle AG, Shiryaev SA, Strongin AY, Tsien RY, Wang Y.\n",
      "PMID: 20197470; PMCID: PMC2840183.\n",
      "View in: PubMed   Mentions: 67     Fields: Neo Neoplasms   Translation:HumansCells\n",
      "Autofluorescent proteins with excitation in the optical window for intravital imaging in mammals. Chem Biol. 2009 Nov 25; 16(11):1169-79.\n",
      "Lin MZ, McKeown MR, Ng HL, Aguilera TA, Shaner NC, Campbell RE, Adams SR, Gross LA, Ma W, Alber T, Tsien RY.\n",
      "PMID: 19942140; PMCID: PMC2814181.\n",
      "View in: PubMed   Mentions: 125     Fields: Bio BiochemistryBio BiologyChe Chemistry   Translation:AnimalsCells\n",
      "Improving the photostability of bright monomeric orange and red fluorescent proteins. Nat Methods. 2008 Jun; 5(6):545-51.\n",
      "Shaner NC, Lin MZ, McKeown MR, Steinbach PA, Hazelwood KL, Davidson MW, Tsien RY.\n",
      "PMID: 18454154; PMCID: PMC2853173.\n",
      "View in: PubMed   Mentions: 475     Fields: Cli Clinical Laboratory Techniques   Translation:AnimalsCells\n",
      "Advances in fluorescent protein technology. J Cell Sci. 2007 Dec 15; 120(Pt 24):4247-60.\n",
      "Shaner NC, Patterson GH, Davidson MW.\n",
      "PMID: 18057027.\n",
      "View in: PubMed   Mentions: 266     Fields: Cel Cell Biology   Translation:AnimalsCells\n",
      "Exploration of new chromophore structures leads to the identification of improved blue fluorescent proteins. Biochemistry. 2007 May 22; 46(20):5904-10.\n",
      "Ai HW, Shaner NC, Cheng Z, Tsien RY, Campbell RE.\n",
      "PMID: 17444659.\n",
      "View in: PubMed   Mentions: 116     Fields: Bio Biochemistry   Translation:Animals\n",
      "Novel chromophores and buried charges control color in mFruits. Biochemistry. 2006 Aug 15; 45(32):9639-47.\n",
      "Shu X, Shaner NC, Yarbrough CA, Tsien RY, Remington SJ.\n",
      "PMID: 16893165.\n",
      "View in: PubMed   Mentions: 210     Fields: Bio Biochemistry   Translation:Animals\n",
      "A guide to choosing fluorescent proteins. Nat Methods. 2005 Dec; 2(12):905-9.\n",
      "Shaner NC, Steinbach PA, Tsien RY.\n",
      "PMID: 16299475.\n",
      "View in: PubMed   Mentions: 1126     Fields: Cli Clinical Laboratory Techniques   Translation:HumansAnimals\n",
      "Dynamics of Z-band based proteins in developing skeletal muscle cells. Cell Motil Cytoskeleton. 2005 May; 61(1):34-48.\n",
      "Wang J, Shaner N, Mittal B, Zhou Q, Chen J, Sanger JM, Sanger JW.\n",
      "PMID: 15810059; PMCID: PMC1993831.\n",
      "View in: PubMed   Mentions: 72     Fields: Cel Cell Biology   Translation:AnimalsCells\n",
      "Actin and alpha-actinin dynamics in the adhesion and motility of EPEC and EHEC on host cells. Cell Motil Cytoskeleton. 2005 Feb; 60(2):104-20.\n",
      "Shaner NC, Sanger JW, Sanger JM.\n",
      "PMID: 15627283.\n",
      "View in: PubMed   Mentions: 11     Fields: Cel Cell Biology   Translation:HumansAnimalsCells\n",
      "Improved monomeric red, orange and yellow fluorescent proteins derived from Discosoma sp. red fluorescent protein. Nat Biotechnol. 2004 Dec; 22(12):1567-72.\n",
      "Shaner NC, Campbell RE, Steinbach PA, Giepmans BN, Palmer AE, Tsien RY.\n",
      "PMID: 15558047.\n",
      "View in: PubMed   Mentions: 2097     Fields: Bio Biotechnology   Translation:AnimalsCells\n",
      "Expression of green or red fluorescent protein (GFP or DsRed) linked proteins in nonmuscle and muscle cells. Mol Biotechnol. 2001 Jan; 17(1):65-71.\n",
      "Ayoob JC, Shaner NC, Sanger JW, Sanger JM.\n",
      "PMID: 11280932.\n",
      "View in: PubMed   Mentions: 6     Fields: Bio BiotechnologyMol Molecular Biology   Translation:AnimalsCells\n",
      "This graph shows the total number of publications by year. To see the data as text, click here.\n",
      "This graph shows the total number of publications by year. To return to the graph, click here.\n",
      "YearPublications200112004120053200612007220081200912010220113201222013220141201532016120173201822019120202202132022120234\n",
      "This graph shows the number and percent of publications by field.\n",
      "Fields are based on how the National Library of Medicine (NLM) classifies the publications' journals and might not represent the specific topics of the publications.\n",
      "Note that an individual publication can be assigned to more than one field. As a result, the publication counts in this graph might add up to more than the number of publications the person has written.\n",
      "To see the data as text, click here.\n",
      "This graph shows the number and percent of publications by field.\n",
      "Fields are based on how the National Library of Medicine (NLM) classifies the publications' journals and might not represent the specific topics of the publications.\n",
      "Note that an individual publication can be assigned to more than one field. As a result, the publication counts in this graph might add up to more than the number of publications the person has written.\n",
      "To see the data as text, click here.\n",
      "Start with:\n",
      "newest\n",
      "oldest\n",
      "Include:\n",
      "line numbers\n",
      "double spacing\n",
      "all authors\n",
      "publication IDs\n",
      "With one click, see people who:\n",
      "Mentor Faculty\n",
      "Run Clinical Trials\n",
      "See more options in the Advanced Search tab\n",
      "Nathan's Networks\n",
      "Concepts\n",
      "(180)\n",
      "Derived automatically from this person's publications.\n",
      "Luminescent Proteins\n",
      "Green Fluorescent Proteins\n",
      "Hydrozoa\n",
      "Abbreviations as Topic\n",
      "Chordata\n",
      "Explore\n",
      "_\n",
      "Co-Authors\n",
      "(9)\n",
      "People in Profiles who have published with this person.\n",
      "Lambert, GerardUCSD\n",
      "Boassa, DanielaUCSD\n",
      "Sprecher, BrittanyUCSD\n",
      "Adams, StephenUCSD\n",
      "Ferro-novick, SusanUCSD\n",
      "Explore\n",
      "_\n",
      "Similar People\n",
      "(60)\n",
      "People who share similar concepts with this person.\n",
      "Hoffman, RobertUCSD\n",
      "Bouvet, MichaelUCSD\n",
      "Gratton, EnricoUCI\n",
      "Looger, LorenUCSD\n",
      "Shu, XiaokunUCSF\n",
      "Explore\n",
      "_\n",
      "Same Department\n",
      "Bayram, Ece\n",
      "Callaway, Edward\n",
      "Cheng, Qun\n",
      "Looger, Loren\n",
      "Powell, Leslie\n",
      "Search Department\n",
      "_\n",
      "UCSD Profiles is managed by the UC San Diego\n",
      "Altman Clinical and Translational Research Institute (ACTRI).\n",
      "This site is running Profiles RNS version UCSF-v3.1.0-2-gbbe5f151 on PROFILES-PWEB03.\n",
      "We use cookies to operate our website. We also use cookies to analyze our site’s performance and improve your experience on our site.\n",
      "For more information about how we use cookies, please see our Website Terms of Use.\n",
      "Home\n",
      "Contact us\n",
      "Copyright © The Regents of the University of California\n"
     ]
    }
   ],
   "source": [
    "#print(scrape_visible_text(\"https://www.bioluminescencehub.org/about-us\"))\n",
    "manual_paste_in = False\n",
    "if manual_paste_in:\n",
    "    text=input(\"text with claims you want validated: \")\n",
    "    source_text=input(\"full source text: \")\n",
    "else:\n",
    "    text, source_text = get_claim_and_source_from_clipboard()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "95165144-41e8-473d-bba2-166af3a9a19e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "source token length: 5649\n",
      "split source into 2 segments\n",
      "encoding\n",
      "0: 0.6294517517089844\n",
      "1: 0.530052661895752\n",
      "[0.6294517517089844, 0.530052661895752]\n",
      "[0]\n",
      "querying openai...\n",
      "\n",
      "\n",
      "\n",
      "QUESTION:\n",
      "\n",
      "    Are the claims made in the following ## Passage supported by the text in the ## Source? Err on the side of \"no\" if you are unsure.\n",
      "    ## Passage\n",
      "    He currently holds the title of Associate Adjunct Professor in Neurosciences at the University of California San Diego.\n",
      "    \n",
      "    ## Source\n",
      "    Nathan Shaner | UCSD Profiles\n",
      "Powered by Altman Clinical and Translational Research Institute\n",
      "UCSD Profiles\n",
      "search, discover, network\n",
      "Home\n",
      "About\n",
      "Help/FAQs\n",
      "History (1)Shaner, NathanSee All Pages\n",
      "Search Options\n",
      "UC Davis\n",
      "UCI\n",
      "UCLA\n",
      "UCSD\n",
      "UCSF\n",
      "All UC Health\n",
      "USC\n",
      "All\n",
      "Everything\n",
      "Research\n",
      "Concepts\n",
      "Awards\n",
      "People\n",
      "UC Health People\n",
      "UC Davis People\n",
      "UCI People\n",
      "UCLA People\n",
      "UCSD People\n",
      "UCSF People\n",
      "USC People\n",
      "Find PeopleFind Everything\n",
      "Login to edit your profile (add a photo, awards, links to other websites, etc.)\n",
      "Edit My Profile\n",
      "My Person List (0)\n",
      "Return to Top\n",
      "Nathan Shaner\n",
      "Title(s)Associate Adjunct Professor, NeurosciencesSchoolVc-health Sciences-schoolsAddress9500 Gilman Drive #La Jolla CA 92093vCardDownload vCard\n",
      "Research  Research Activities and FundingHighly specific control of neurons with photoswitchable bioluminescent optogenetics.NIH R21EY030716Aug 1, 2019 - Jul 31, 2021Role: Principal InvestigatorProbes for Luminescence-based Superresolution MicroscopyNIH R01GM121944Feb 16, 2017 - Jan 31, 2021Role: Principal InvestigatorBioLuminescent OptoGenetics (BL-OG): A Novel and Versatile Strategy for NeuromodulationNIH U01NS099709Sep 30, 2016 - Jun 30, 2020Role: Co-Principal InvestigatorFluorescent proteins for superresolution imagingNIH R01GM109984Sep 30, 2014 - Aug 31, 2020Role: Principal Investigator\n",
      "Bibliographic  Publications\n",
      "Publications listed below are automatically derived from MEDLINE/PubMed and other sources, which might result in incorrect or missing publications.\n",
      "Researchers can login to make corrections and additions, or contact us for help.\n",
      "to make corrections and additions.\n",
      "Newest\n",
      "|\n",
      "Oldest\n",
      "|\n",
      "Most Cited\n",
      "|\n",
      "Most Discussed\n",
      "|\n",
      "Timeline\n",
      "|\n",
      "Field Summary\n",
      "|\n",
      "Plain Text\n",
      "Altmetrics Details\n",
      "PMC Citations indicate the number of times the publication was cited by articles in PubMed Central, and the Altmetric score represents citations in news articles and social media.\n",
      "(Note that publications are often cited in additional ways that are not shown here.)\n",
      "Fields are based on how the National Library of Medicine (NLM) classifies the publication's journal and might not represent the specific topic of the publication.\n",
      "Translation tags are based on the publication type and the MeSH terms NLM assigns to the publication.\n",
      "Some publications (especially newer ones and publications not in PubMed) might not yet be assigned Field or Translation tags.)\n",
      "Click a Field or Translation tag to filter the publications.\n",
      "Bioluminescent Genetically Encoded Glutamate Indicators for Molecular Imaging of Neuronal Activity. ACS Synth Biol. 2023 08 18; 12(8):2301-2309.\n",
      "Petersen ED, Lapan AP, Castellanos Franco EA, Fillion AJ, Crespo EL, Lambert GG, Grady CJ, Zanca AT, Orcutt R, Hochgeschwender U, Shaner NC, Gilad AA.\n",
      "PMID: 37450884; PMCID: PMC10443529.\n",
      "View in: PubMed   Mentions: 1     Fields: Bio BiotechnologyMol Molecular Biology   Translation:Humans\n",
      "A Bioluminescent Activity Dependent (BLADe) Platform for Converting Neuronal Activity to Photoreceptor Activation. bioRxiv. 2023 Jun 27.\n",
      "Crespo EL, Pal A, Prakash M, Silvagnoli AD, Zaidi Z, Gomez-Ramirez M, Tree MO, Shaner NC, Lipscombe D, Moore C, Hochgeschwender U.\n",
      "PMID: 37425742; PMCID: PMC10327117.\n",
      "View in: PubMed   Mentions:\n",
      "CaBLAM! A high-contrast bioluminescent Ca 2+ indicator derived from an engineered Oplophorus gracilirostris luciferase. bioRxiv. 2023 Jun 26.\n",
      "Lambert GG, Crespo EL, Murphy J, Boassa D, Luong S, Hu J, Sprecher B, Tree MO, Orcutt R, Heydari D, Bell AB, Torreblanca Zanca A, Hakimi A, Lipscombe D, Moore CI, Hochgeschwender U, Shaner NC.\n",
      "PMID: 37425712; PMCID: PMC10327125.\n",
      "View in: PubMed   Mentions:\n",
      "A FRET-based excitatory luminopsin with improved efficacy of bioluminescence-driven channelrhodopsin activation. bioRxiv. 2023 Jun 26.\n",
      "Björefeldt A, Murphy J, Crespo EL, Prakash M, Friedman N, Brown TM, Lipscombe D, Moore CI, Hochgeschwender U, Shaner NC.\n",
      "PMID: 37425735; PMCID: PMC10327108.\n",
      "View in: PubMed   Mentions:\n",
      "Selective control of synaptically-connected circuit elements by all-optical synapses. Commun Biol. 2022 01 11; 5(1):33.\n",
      "Prakash M, Murphy J, St Laurent R, Friedman N, Crespo EL, Bjorefeldt A, Pal A, Bhagat Y, Kauer JA, Shaner NC, Lipscombe D, Moore CI, Hochgeschwender U.\n",
      "PMID: 35017641; PMCID: PMC8752598.\n",
      "View in: PubMed   Mentions: 7  Translation:AnimalsCells\n",
      "Endoplasmic reticulum tubules limit the size of misfolded protein condensates. Elife. 2021 09 01; 10.\n",
      "Parashar S, Chidambaram R, Chen S, Liem CR, Griffis E, Lambert GG, Shaner NC, Wortham M, Hay JC, Ferro-Novick S.\n",
      "PMID: 34467852; PMCID: PMC8486381.\n",
      "View in: PubMed   Mentions: 10     Fields: Bio Biology   Translation:HumansAnimalsCells\n",
      "Bioluminescent optogenetic (BL-OG) activation of neurons during mouse postnatal brain development. STAR Protoc. 2021 09 17; 2(3):100667.\n",
      "Crespo EL, Prakash M, Bjorefeldt A, Medendorp WE, Shaner NC, Lipscombe D, Moore CI, Hochgeschwender U.\n",
      "PMID: 34286295; PMCID: PMC8273415.\n",
      "View in: PubMed   Mentions: 2     Fields: Bio BiologySci Science   Translation:AnimalsCells\n",
      "Local Electric Field Controls Fluorescence Quantum Yield of Red and Far-Red Fluorescent Proteins. Front Mol Biosci. 2021; 8:633217.\n",
      "Drobizhev M, Molina RS, Callis PR, Scott JN, Lambert GG, Salih A, Shaner NC, Hughes TE.\n",
      "PMID: 33763453; PMCID: PMC7983054.\n",
      "View in: PubMed   Mentions: 9\n",
      "Aequorea's secrets revealed: New fluorescent proteins with unique properties for bioimaging and biosensing. PLoS Biol. 2020 11; 18(11):e3000936.\n",
      "Lambert GG, Depernet H, Gotthard G, Schultz DT, Navizet I, Lambert T, Adams SR, Torreblanca-Zanca A, Chu M, Bindels DS, Levesque V, Nero Moffatt J, Salih A, Royant A, Shaner NC.\n",
      "PMID: 33137097; PMCID: PMC7660908.\n",
      "View in: PubMed   Mentions: 15     Fields: Bio Biology   Translation:Animals\n",
      "Miniaturized Devices for Bioluminescence Imaging in Freely Behaving Animals. Annu Int Conf IEEE Eng Med Biol Soc. 2020 07; 2020:4385-4389.\n",
      "Celinskis D, Friedman N, Koksharov M, Murphy J, Gomez-Ramirez M, Borton D, Shaner N, Hochgeschwender U, Lipscombe D, Moore C.\n",
      "PMID: 33018967.\n",
      "View in: PubMed   Mentions: 2     Fields: Bio Biomedical Engineering   Translation:AnimalsCells\n",
      "Alkynamide phthalazinones as a new class of TbrPDEB1 inhibitors (Part 2). Bioorg Med Chem. 2019 09 15; 27(18):4013-4029.\n",
      "de Heuvel E, Singh AK, Boronat P, Kooistra AJ, van der Meer T, Sadek P, Blaazer AR, Shaner NC, Bindels DS, Caljon G, Maes L, Sterk GJ, Siderius M, Oberholzer M, de Esch IJP, Brown DG, Leurs R.\n",
      "PMID: 31378593.\n",
      "View in: PubMed   Mentions: 6     Fields: Bio BiochemistryChe Chemistry   Translation:Humans\n",
      "Assay for Detecting Gαi-Mediated Decreases in cAMP in Living Cells. SLAS Discov. 2018 10; 23(9):898-906.\n",
      "Tewson P, Martinka S, Shaner N, Berlot C, Quinn AM, Hughes T.\n",
      "PMID: 29991302.\n",
      "View in: PubMed   Mentions: 6     Fields: Mol Molecular Biology   Translation:HumansAnimalsCells\n",
      "Spatiotemporal manipulation of ciliary glutamylation reveals its roles in intraciliary trafficking and Hedgehog signaling. Nat Commun. 2018 04 30; 9(1):1732.\n",
      "Hong SR, Wang CL, Huang YS, Chang YC, Chang YC, Pusapati GV, Lin CY, Hsu N, Cheng HC, Chiang YC, Huang WE, Shaner NC, Rohatgi R, Inoue T, Lin YC.\n",
      "PMID: 29712905; PMCID: PMC5928066.\n",
      "View in: PubMed   Mentions: 33     Fields: Bio BiologySci Science   Translation:AnimalsCells\n",
      "Characterization of a spectrally diverse set of fluorescent proteins as FRET acceptors for mTurquoise2. Sci Rep. 2017 09 20; 7(1):11999.\n",
      "Mastop M, Bindels DS, Shaner NC, Postma M, Gadella TWJ, Goedhart J.\n",
      "PMID: 28931898; PMCID: PMC5607329.\n",
      "View in: PubMed   Mentions: 35     Fields: Sci Science   Translation:HumansAnimalsCells\n",
      "Blue-Shifted Green Fluorescent Protein Homologues Are Brighter than Enhanced Green Fluorescent Protein under Two-Photon Excitation. J Phys Chem Lett. 2017 Jun 15; 8(12):2548-2554.\n",
      "Molina RS, Tran TM, Campbell RE, Lambert GG, Salih A, Shaner NC, Hughes TE, Drobizhev M.\n",
      "PMID: 28530831; PMCID: PMC5474692.\n",
      "View in: PubMed   Mentions: 16     Fields: Che Chemistry\n",
      "Engineering of mCherry variants with long Stokes shift, red-shifted fluorescence, and low cytotoxicity. PLoS One. 2017; 12(2):e0171257.\n",
      "Shen Y, Chen Y, Wu J, Shaner NC, Campbell RE.\n",
      "PMID: 28241009; PMCID: PMC5328254.\n",
      "View in: PubMed   Mentions: 28     Fields: Med Medicine (General)Sci Science   Translation:AnimalsCells\n",
      "Structural analysis of the bright monomeric yellow-green fluorescent protein mNeonGreen obtained by directed evolution. Acta Crystallogr D Struct Biol. 2016 12 01; 72(Pt 12):1298-1307.\n",
      "Clavel D, Gotthard G, von Stetten D, De Sanctis D, Pasquier H, Lambert GG, Shaner NC, Royant A.\n",
      "PMID: 27917830; PMCID: PMC5137226.\n",
      "View in: PubMed   Mentions: 15  Translation:AnimalsCells\n",
      "New DAG and cAMP Sensors Optimized for Live-Cell Assays in Automated Laboratories. J Biomol Screen. 2016 Mar; 21(3):298-305.\n",
      "Tewson PH, Martinka S, Shaner NC, Hughes TE, Quinn AM.\n",
      "PMID: 26657040; PMCID: PMC4766961.\n",
      "View in: PubMed   Mentions: 53     Fields: Mol Molecular Biology   Translation:HumansCells\n",
      "Special Section Guest Editorial: Protein Photonics for Imaging, Sensing, and Manipulation: Honoring Prof. Osamu Shimomura, a Pioneer of Photonics for Biomedical Research. J Biomed Opt. 2015 Oct; 20(10):101201.\n",
      "Fujita K, Nagai T, Shaner N, Egner A.\n",
      "PMID: 26527157.\n",
      "View in: PubMed   Mentions:    Fields: Bio Biomedical EngineeringOph Ophthalmology   Translation:HumansAnimals\n",
      "Occurrence of Isopenicillin-N-Synthase Homologs in Bioluminescent Ctenophores and Implications for Coelenterazine Biosynthesis. PLoS One. 2015; 10(6):e0128742.\n",
      "Francis WR, Shaner NC, Christianson LM, Powers ML, Haddock SH.\n",
      "PMID: 26125183; PMCID: PMC4488382.\n",
      "View in: PubMed   Mentions: 11     Fields: Med Medicine (General)Sci Science   Translation:AnimalsCells\n",
      "Fluorescent proteins for quantitative microscopy: important properties and practical evaluation. Methods Cell Biol. 2014; 123:95-111.\n",
      "Shaner NC.\n",
      "PMID: 24974024.\n",
      "View in: PubMed   Mentions: 13     Fields: Cel Cell Biology   Translation:HumansAnimals\n",
      "A bright monomeric green fluorescent protein derived from Branchiostoma lanceolatum. Nat Methods. 2013 May; 10(5):407-9.\n",
      "Shaner NC, Lambert GG, Chammas A, Ni Y, Cranfill PJ, Baird MA, Sell BR, Allen JR, Day RN, Israelsson M, Davidson MW, Wang J.\n",
      "PMID: 23524392; PMCID: PMC3811051.\n",
      "View in: PubMed   Mentions: 568     Fields: Cli Clinical Laboratory Techniques   Translation:Animals\n",
      "A comparison across non-model animals suggests an optimal sequencing depth for de novo transcriptome assembly. BMC Genomics. 2013 Mar 12; 14:167.\n",
      "Francis WR, Christianson LM, Kiko R, Powers ML, Shaner NC, Haddock SH.\n",
      "PMID: 23496952; PMCID: PMC3655071.\n",
      "View in: PubMed   Mentions: 48     Fields: Gen Genetics   Translation:AnimalsCells\n",
      "Expression and characterization of the calcium-activated photoprotein from the ctenophore Bathocyroe fosteri: insights into light-sensitive photoproteins. Biochem Biophys Res Commun. 2013 Feb 08; 431(2):360-6.\n",
      "Powers ML, McDermott AG, Shaner NC, Haddock SH.\n",
      "PMID: 23262181; PMCID: PMC3570696.\n",
      "View in: PubMed   Mentions: 3     Fields: Bio BiochemistryBio Biophysics   Translation:AnimalsCells\n",
      "The mFruit collection of monomeric fluorescent proteins. Clin Chem. 2013 Feb; 59(2):440-1.\n",
      "Shaner NC.\n",
      "PMID: 23213078.\n",
      "View in: PubMed   Mentions: 1     Fields: Che Chemistry   Translation:Animals\n",
      "Concurrent imaging of synaptic ves\n",
      "    \n",
      "RESPONSE:\n",
      "The relevant quotes from the ## source are:\n",
      "\n",
      "- \"Nathan Shaner\n",
      "Title(s)Associate Adjunct Professor, NeurosciencesSchoolVc-health Sciences-schoolsAddress9500 Gilman Drive #La Jolla CA 92093vCardDownload vCard\"\n",
      "\n",
      "Based on this information, the ## passage is supported by the ## source. The ## source confirms that Nathan Shaner holds the title of Associate Adjunct Professor in Neurosciences at the University of California San Diego.\n",
      "\n",
      "\n",
      "\n",
      "Passage being validated:\n",
      "He currently holds the title of Associate Adjunct Professor in Neurosciences at the University of California San Diego.\n",
      "\n",
      "\n",
      "Answer: \n",
      "The relevant quotes from the ## source are:\n",
      "\n",
      "- \"Nathan Shaner\n",
      "Title(s)Associate Adjunct Professor, NeurosciencesSchoolVc-health Sciences-schoolsAddress9500 Gilman Drive #La Jolla CA 92093vCardDownload vCard\"\n",
      "\n",
      "Based on this information, the ## passage is supported by the ## source. The ## source confirms that Nathan Shaner holds the title of Associate Adjunct Professor in Neurosciences at the University of California San Diego.\n"
     ]
    }
   ],
   "source": [
    "model = \"gpt-3.5-turbo\" #\"gpt-4\"\n",
    "sys_message = \"Your task is to determine if a ## passage pulled from Wikipedia is factually supported by its ## source. the ## source is ALL of the text after '## Source'. most information in the ## source will not be relevant to your task, so you must first find and print out quotes of the ## source that support or refute the ## Passage. Only print out quotes from the ## source that are relevant to determining the validity of the ## passage. then, briefly and concisely use the relevant quotes from the ## source to decide if the ## passage is supported by the ## source.\"\n",
    "\n",
    "#splitting up source if it is too big\n",
    "enc = encoding_for_model(model)\n",
    "toks = enc.encode(source_text)\n",
    "print(\"source token length: \"+ str(len(toks)))\n",
    "question = f\"\"\"\n",
    "Are the claims made in the following ## Passage supported by the text in the ## Source?\n",
    "\n",
    "## Passage\n",
    "{text}\n",
    "\n",
    "## Source\n",
    "\n",
    "\n",
    "\"\"\"\n",
    "basemessagetoks = enc.encode(sys_message+ question)\n",
    "#print(\"base message token length: \"+ str(len(basemessagetoks)))\n",
    "minimum_response_length_tokens = 400\n",
    "maxlength = 4097-15-minimum_response_length_tokens\n",
    "if model == \"gpt-4\":\n",
    "    maxlength = 8192-15-minimum_response_length_tokens\n",
    "source_max_length = maxlength - len(basemessagetoks)\n",
    "#print(\"max source token length: \" + str(source_max_length))\n",
    "textsegments = []\n",
    "for i in range(0,len(toks),source_max_length-15):\n",
    "    textsegments.append( \"\".join([enc.decode_single_token_bytes(o).decode('utf-8', 'ignore') for o in toks[i:min(i+source_max_length, len(toks))]]) )\n",
    "\n",
    "print(\"split source into \" + str(len(textsegments)) + \" segments\")\n",
    "\n",
    "#find top N most relevant segments of source via cos similarity\n",
    "send_at_most_N_gpt_source_requests = 1\n",
    "if len(textsegments)>send_at_most_N_gpt_source_requests:\n",
    "    emb_model = SentenceTransformer(\"BAAI/bge-small-en-v1.5\", device=0)\n",
    "    print(\"encoding\")\n",
    "    x = emb_model.encode([text] + textsegments, convert_to_tensor=True)\n",
    "    cos_similarities = []\n",
    "    for seg in range(0,len(textsegments)):\n",
    "        cos_similarities.append(float(F.cosine_similarity(x[0], x[seg+1], dim=0).item()))\n",
    "        print(str(seg)+ \": \" + str(cos_similarities[seg]))\n",
    "    \n",
    "    #indices_of_largest = np.argpartition(cos_similarities, send_at_most_N_gpt_source_requests)[:send_at_most_N_gpt_source_requests]\n",
    "    print(cos_similarities)\n",
    "    indices_of_largest = sorted(range(len(cos_similarities)), key=lambda sub: cos_similarities[sub])[-send_at_most_N_gpt_source_requests:]\n",
    "\n",
    "    print(indices_of_largest)\n",
    "    top_text_segments = [textsegments[i] for i in indices_of_largest]\n",
    "    textsegments=top_text_segments\n",
    "\n",
    "###FINALLY QUERYING CHATGPT\n",
    "results = []\n",
    "for segment in textsegments:\n",
    "    question = f\"\"\"\n",
    "    Are the claims made in the following ## Passage supported by the text in the ## Source? Err on the side of \"no\" if you are unsure.\n",
    "    ## Passage\n",
    "    {text}\n",
    "    \n",
    "    ## Source\n",
    "    {segment}\n",
    "    \"\"\"\n",
    "    #print(question)\n",
    "    print(\"querying openai...\")\n",
    "    c = openai.ChatCompletion.create(\n",
    "        model=model,\n",
    "        messages=[{\"role\": \"system\", \"content\": sys_message},\n",
    "                  {\"role\": \"user\", \"content\": question}])\n",
    "    results.append(c['choices'][0]['message']['content'])\n",
    "    if verbose:\n",
    "        print(\"\\n\\n\\nQUESTION:\")\n",
    "        print(question)\n",
    "        print(\"RESPONSE:\")\n",
    "        print(c['choices'][0]['message']['content'])\n",
    "#extra work if too big for context length\n",
    "if len(results) >1:\n",
    "    sourcenum = 1\n",
    "    sources = \"\"\n",
    "    for result in results:\n",
    "        sources += \"\\nAnalysis part #\" + str(sourcenum) + \" of source\\n \"\n",
    "        sources += result\n",
    "        sourcenum=sourcenum+1\n",
    "    \n",
    "    \n",
    "    question = f\"\"\"\n",
    "        Are the claims made in the following ## Passage supported by evidence in any part of the text after ## Source? simply answer yes, no, or unknown, and provide direct quotes from relevant sections of the ##source as breif evidence.\n",
    "        ## Passage\n",
    "        {text}\n",
    "        \n",
    "        ## Source\n",
    "        {sources}\n",
    "    \"\"\"\n",
    "    c = openai.ChatCompletion.create(\n",
    "        model=model,\n",
    "        messages=[{\"role\": \"system\", \"content\": sys_message},\n",
    "                    {\"role\": \"user\", \"content\": question}])\n",
    "    print(\"answer: \")\n",
    "    print(c['choices'][0]['message']['content'])\n",
    "else:\n",
    "    print(\"\\n\\n\\nPassage being validated:\\n\" + text)\n",
    "    print(\"\\n\\nAnswer: \")\n",
    "    print(results[0])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "455fda77-5489-4030-b207-6be94ade5b9b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The relevant quotes from the ## source are:\n",
      "\n",
      "- \"Nathan Shaner\n",
      "Title(s)Associate Adjunct Professor, NeurosciencesSchoolVc-health Sciences-schoolsAddress9500 Gilman Drive #La Jolla CA 92093vCardDownload vCard\"\n",
      "\n",
      "Based on this information, the ## passage is supported by the ## source. The ## source confirms that Nathan Shaner holds the title of Associate Adjunct Professor in Neurosciences at the University of California San Diego.\n"
     ]
    }
   ],
   "source": [
    "print(results[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5b47f14b-555d-49cb-87d7-96c34abcc619",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
